### ISSUE TITLE
Exploding gradients when using both backward() and DataParallel

### ISSUE THREAD

**Issue by ErikStammes** [pytorch/pytorch#31768]
## 🐛 Bug

I am building a model that needs the gradients computed during a backward pass, in the forward of my network. To speed up experimentation I am using nn.DataParallel to spread the workload across multiple GPUs. However, I’m running into a problem of exploding gradients (and losses) when I compute the backward() from the forward() function when I use multiple GPUs.

The following code is a copy of the CIFAR-10 tutorial, except that the forward() function now also does a backward to collect gradients (I would usually do this using backward hooks, but have left out that piece of code for clarity), I have also wrapped the network inside nn.DataParallel

## To Reproduce

To reproduce the behavior:

```import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=4,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)

        self.zero_grad()
        predictions = F.softmax(x).argmax(dim=1)
        ohe = torch.zeros_like(x)
        for i, label in enumerate(predictions):
            ohe[i, label] = 1

        ohe = torch.autograd.Variable(ohe)
        x.backward(gradient=ohe, retain_graph=True)
        self.zero_grad()

        return x


net = Net()
net.cuda()
if torch.cuda.device_count() > 1:
    net = nn.DataParallel(net)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

for epoch in range(2):  # loop over the dataset multiple times

    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data
        inputs, labels = inputs.cuda(), labels.cuda()

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        if i % 2000 == 1999:    # print every 2000 mini-batches
            print('[%d, %5d] loss: %.3f' %
                  (epoch + 1, i + 1, running_loss / 2000))
            running_loss = 0.0

print('Finished Training')
```

## Expected behavior

When I run this code on a single GPU, it works fine (i.e. losses stay in normal range), when I run it on multiple GPUs, the losses explode.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.5 LTS
GCC version: (GCC) 9.2.0
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti
GPU 4: GeForce GTX 1080 Ti
GPU 5: GeForce GTX 1080 Ti

Nvidia driver version: 418.56
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect

cc @ezyang @gchanan @zou3519

**Comment by ezyang** [comment]
Marking as high priority to attempt reproducing using the provided script

**Comment by ezyang** [comment]
@pietern points out that the math is different when you use DataParallel, so it may not actually be DataParallel. This stuff is notoriously tricky. But there is no BatchNorm, so there shouldn't really be different behavior, says @ngimel

**Comment by albanD** [comment]
Hi,

Does moving the `zero_grad()` in the main loop after the forward pass solve the issue (just before the loss.backward())?

Also using `.backward()` in your forward pass is quite dangerous, you don't know how much it will impact as it will backprop on anything that impacted `x` which does not seem to be what you want here right? Could you use `autograd.grad()` to get only the gradient you want?

**Comment by peterbell10** [comment]
>Does moving the zero_grad() in the main loop after the forward pass solve the issue (just before the loss.backward())?

It does.

**Comment by ErikStammes** [comment]
> Does moving the zero_grad() in the main loop after the forward pass solve the issue (just before the loss.backward())?
That indeed solves the issue, thanks!
